### DISCUSSION TITLE
[data-importer] Could not create transaction. Unexpected empty response from Firefly III. Check the logs.

### DISCUSSION THREAD

**Discussion by soxslayer** [firefly-iii/firefly-iii#6055] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I'm trying to import transactions into Firefly iii using the data importer. Both Firefly iii and the data importer are running in their respective docker images. After importing, the data importer reports:

`Could not create transaction. Unexpected empty response from Firefly III. Check the logs.`

This issue has started happening sometime between now and last week. I tried reverting back the last known working versions of the docker images, `5.7.2` for core and `0.9.8` for the data importer, and still see the same issue.

### Expected behaviour

The data importer should successfully import the transactions specified in the CSV file.

**Note**
I am able to successfully create transactions manually through the web UI.

### Debug information

Debug information generated at 2022-05-03 20:05:47 America/Denver for Firefly III version **5.7.4**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.7.4 |
| Firefly III API | 1.5.6 |
| Build | 174, 02-05-2022 19:49:26 CEST |
| DB version | 18 (exp. 18) |
| PHP | `8.0.17` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | America/Denver |
| Browser TZ | America/Denver |
| App environment | local |
| App debug mode | true |
| App cache driver | file |
| App logging | debug, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 24 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `**` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2022-05-01 00:00:00 |
| Session end | 2022-05-31 23:59:59 |
| Session first | 2021-09-27 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:97.0) Gecko/20100101 Firefox/97.0 |

### Steps to reproduce

Unfortunately, I have been using my self-hosted instance of Firefly iii for a couple of weeks now and cannot pinpoint steps to recreate my setup. Last week I imported my transactions and everything function correctly. Yesterday when I tried to import transactions this error started occurring.

### Additional info

Here are, what seem to be, the relevant logs leading up to this error, sensitive data redacted. The transaction it's attempting to submit to Firefly iii matches the transaction that I have defined in my CSV file.

```
importer_1  | [2022-05-04 04:47:51] production.DEBUG: Submitting to Firefly III: {"group_title":null,"error_if_duplicate_hash":true,"transactions":[{"type":"withdrawal","date":"2022-05-01 00:00:00","currency_id":8,"currency_code":null,"amount":"111.970000000000","description":"xxx","source_id":26,"source_name":null,"destination_id":null,"destination_name":"(no name)","process_date":"2022-05-02 00:00:00","original_source":"jc5-data-import-v0.9.9","tags":[],"source_iban":null,"source_number":null,"source_bic":null,"destination_iban":null,"destination_number":null,"destination_bic":null}]} 
app_1       | [2022-05-03 20:47:51] local.DEBUG: GroupCollector: The limit is now 50  
app_1       | [2022-05-03 20:47:51] local.DEBUG: GroupCollector: page is now 1  
importer_1  | [2022-05-04 04:47:51] production.DEBUG: isReadyForFileStep("submit")  
importer_1  | [2022-05-04 04:47:51] production.DEBUG: isReadyForFileStep: Return true  
importer_1  | [2022-05-04 04:47:51] production.DEBUG: Now at App\Http\Controllers\Import\SubmitController::status(conv-ObJuG3KruUQP)  
importer_1  | [2022-05-04 04:47:51] production.DEBUG: Now in startOrFindJob(conv-ObJuG3KruUQP)  
importer_1  | [2022-05-04 04:47:51] production.DEBUG: Try to see if file exists for job conv-ObJuG3KruUQP.  
importer_1  | [2022-05-04 04:47:51] production.DEBUG: Status file exists for job conv-ObJuG3KruUQP.  
importer_1  | 10.4.86.4 - - [04/May/2022:02:47:51 +0000] "GET /import/submit/status?identifier=conv-ObJuG3KruUQP HTTP/1.1" 200 1180 "http://xxx/import/submit" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:97.0) Gecko/20100101 Firefox/97.0"
app_1       | 192.168.112.5 - - [03/May/2022:20:47:51 -0600] "GET /api/v1/transactions HTTP/1.0" 200 91488 "-" "GuzzleHttp/7"
```

According to the data importer logs the raw response from the above Firefly iii request was the following, sensitive data redacted:

```
importer_1  | [2022-05-04 04:47:52] production.ERROR: Could not create transaction. Unexpected empty response from Firefly III. Check the logs. [{"type":"transactions","id":"786","attributes":{"created_at":"2022-04-25T22:32:02-06:00","updated_at":"2022-04-25T22:32:02-06:00","user":"1","group_title":null,"transactions":[{"user":"1","transaction_journal_id":"788","type":"withdrawal","date":"2022-04-25T00:00:00-06:00","order":0,"currency_id":"8","currency_code":"USD","currency_name":"US Dollar","currency_symbol":"$","currency_decimal_places":2,"foreign_currency_id":"0","foreign_currency_code":null,"foreign_currency_symbol":null,"foreign_currency_decimal_places":0,"amount":"15.950000000000000000000000","foreign_amount":null,"description":"xxxx","source_id":"27","source_name":"xxxx","source_iban":"","source_type":"Asset account","destination_id":"31","destination_name":"(no name)","destination_iban":null,"destination_type":"Expense account","budget_id":"7","budget_name":"xxx","category_id":"10","category_name":"xxx","bill_id":null,"bill_name":null,"reconciled":false,"notes":null,"tags":["Data Import on 2022-04-26 @ 06:31"],"internal_reference":null,"external_id":"xxx","original_source":"ff3-v5.7.2|api-v1.5.6","recurrence_id":null,"recurrence_total":null,"recurrence_count":null,"bunq_payment_id":null,"external_url":null,"import_hash_v2":"ca9ec261cabc65802d34eaf0671fd2751245f7fd7f4341c83fdb88dc98a79b31","sepa_cc":null,"sepa_ct_op":null,"sepa_ct_id":null,"sepa_db":null,"sepa_country":null,"sepa_ep":null,"sepa_ci":null,"sepa_batch_id":null,"interest_date":null,"book_date":null,"process_date":null,"due_date":null,"payment_date":null,"invoice_date":null,"longitude":null,"latitude":null,"zoom_level":null,"has_attachments":false}]},"links":{"self":"https://xxx/api/v1/transactions/786","0":{"rel":"self","uri":"/transactions/786"}}}, <49 other transactions>]
```

The transactions that are listed above are old transactions from previous imports.

Looking into the code, the error seems to originate from:

`./Services/Shared/Import/Routine/ApiSubmitter.php:286`

```
$group = $response->getTransactionGroup();
if (null === $group) {
    $message = 'Could not create transaction. Unexpected empty response from Firefly III. Check the logs.';
    app('log')->error($message, $response->getRawData());
    $this->addError($index, $message);

    return $return;
}
```

Looking into the source code for `PostTransactionResponse` it seems to expect a single transaction-like object:

```
class PostTransactionResponse extends Response
{
   <some lines of code here>
   public function __construct(array $data)
    {
        $this->transactionGroup = null;
        if (isset($data['id'])) {
            $this->transactionGroup = TransactionGroup::fromArray($data);
        }
        $this->rawData = $data;
    }
```

but according to the log the API request to Firefly iii is returning an array of transactions. All other functions within the Firefly iii user interface seem to be functioning correctly.

Given I don't have concrete steps to get Firefly iii into this state for repro, I'm more than happy to do as much debugging on my end as necessary. I don't have a background in php nor webapp development so I may need super simplified instructions or pointers on topics to research if action is needed on my end.

**Comment by JC5** [comment]
Please update to the latest version of the data importer, and Firefly III, and make sure you update all composer packages as well.

**Reply by soxslayer** [reply]
I am currently using the `lastest` tag of the docker images, both `core` and `data-importer`. I did an explicit `docker compose pull` to update this morning and still see the same behavior. I'm not quite sure what you mean by "composer packages" but I'm assuming you mean the [Firefly iii docker GitHub repo](https://github.com/firefly-iii/docker). I didn't directly clone that entire repo, but I did use the `docker-compose.yaml` file as the basis for mine.

Here is my compose file if this helps:

```
version: '3'

services:
  db:
    image: mariadb
    restart: always
    volumes:
      - db:/var/lib/mysql
    env_file:
      - db.env
      - db_root.env

  app:
    image: fireflyiii/core:latest
    restart: always
    volumes:
      - firefly:/var/www/html/storage/upload
    env_file: app.env
    depends_on:
      - db

  importer:
    image: fireflyiii/data-importer:latest
    env_file: fidi.env
    ports:
      - "5555:8080"

volumes:
  db:
  firefly:
```

I have also updated my system to use the new built-in `docker-compose-plugins` instead of `docker-compose`.

Additionally, I tried the `develop` tag for the `core` and `data-importer` docker images. All of the above resulted in the same error and same information in the log files.

If there is anything that I have missed in updating please let me know.

**Reply by JC5** [reply]
Could you share with me over email a CSV file with a transaction that reproduces the issue? Also please your JSON config file.

**Reply by soxslayer** [reply]
Sent. Your support is much appreciated.

**Reply by JC5** [reply]
I'm sorry, but it seems to work for me ™️ .

Tracing the FIDI logs I see:

```
[2022-05-04 17:52:12] local.DEBUG: Add message on index #0 (line no. 1): Created withdrawal <a target="_blank" href="http://firefly.jc.home/transactions/show/225">#225 "EUR ******"</a> (EUR ******)
```

**Reply by soxslayer** [reply]
I had a feeling that it was something specific to my particular instance. Since this is only happening on my end I'm willing to dive into this and try and debug from my end. Like I mentioned, I'm not too familiar with php and webapp development. Is there any documentation on setting up a development environment specific to Firefly iii? I couldn't find anything in the user guide. Do you have any suggestions on where / how to start in the code base?

**Reply by JC5** [reply]
I'll put something on paper over the weekend!

**Reply by soxslayer** [reply]
I was able to do a database dump and import into a new set of docker images / volumes and the problem still persists. I can provide a sanitized dump of my database so you can try and recreate this problem on your end.

If you'd like to try and recreate this issue. What all would you require? Just the database dump?

**Reply by JC5** [reply]
I would like to have 1 transaction in a csv file that reproduces the issue and the json configuration file.

**Reply by soxslayer** [reply]
Right, since it appears that the current state of my database is a factor in this issue, I can provide you with a dump of my database, 1 CSV transaction file and a JSON configuration to recreate this issue. Is this acceptable?

**Reply by JC5** [reply]
I don't need the database, I should be able to recreate the issue with the transaction itself.

**Reply by soxslayer** [reply]
Then I'm back at square one. The transaction that's failing for me is the one I emailed to you earlier in the week. You said you could not reproduce the failure on your end with the transaction. So it only seems to be failing when importing into my database.

**Reply by JC5** [reply]
Sorry I didn't realise, send over an SQL dump by all means

**Reply by soxslayer** [reply]
Sent via email.

**Reply by JC5** [reply]
Sorry man, it still worked. I did however find some other issues with the database integrity checks that Firefly executes. Your DB has some things that these scripts don't catch.

Other than that, I'm running out of ideas :(

**Reply by soxslayer** [reply]
Could you please describe the setup you're using? Every time I've tried with that database dump and the transaction with my setup I was getting the failure. I'm wondering if my docker images are somehow older or something else with my setup is amuck.

**Reply by JC5** [reply]
I created a new data importer + Firefly III Docker installation using this Docker file and .env file: https://github.com/firefly-iii/docker/blob/main/docker-compose-data.yml

The only thing I changed, I added port 82 to the database so i could reach it using Querious.

I imported your DB, and reset the password using the reset function. I had to set the logging to debug.

Using the account and reset pw, I created a client ID (non-confidential client) at /profile and used it on the data importer. ID was 136.

I then uploaded your unchanged CSV and JSON. And I got one transaction imported succesfully.

**Reply by soxslayer** [reply]
So, I tried pointing my data-importer to the Docker DNS name, instead of my public URL, and everything seems to be working again.

**Reply by JC5** [reply]
That's weird! Although I'm glad it works now.
